File: frontend/test/features/tenant_facility/domain/facility_similarity_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:hosspi_hms/core/permissions/access_policy.dart';
import 'package:hosspi_hms/core/security/auth_session.dart';
import 'package:hosspi_hms/core/security/session_tokens.dart';
import 'package:hosspi_hms/features/tenant_facility/domain/entities/facility_similarity.dart';
import 'package:hosspi_hms/features/tenant_facility/domain/entities/tenant_facility_setup.dart';

void main() {
  group('checkFacilityDuplicates', () {
    const List<FacilityProfile> existing = <FacilityProfile>[
      FacilityProfile(
        id: 'FAC0001',
        tenantId: 'TEN0001',
        name: 'DemoCare General Hospital',
        type: FacilitySetupType.hospital,
        isActive: true,
        phone: '+256700000000',
        email: 'info@democare.test',
        addressLine1: '12 Kampala Road',
        city: 'Kampala',
        country: 'Uganda',
      ),
      FacilityProfile(
        id: 'FAC0002',
        tenantId: 'TEN0001',
        name: 'Acme Clinic',
        type: FacilitySetupType.clinic,
        isActive: false,
      ),
    ];

    test('detects exact name conflict within tenant', () {
      final FacilityDuplicateCheckResult result = checkFacilityDuplicates(
        name: 'DemoCare General Hospital',
        type: FacilitySetupType.hospital,
        isActive: true,
        existing: existing,
      );

      expect(result.exactNameConflict, isTrue);
      expect(result.similarMatches, isNotEmpty);
      expect(result.overridableMatches, isEmpty);
    });

    test('detects similar facility names above threshold', () {
      final FacilityDuplicateCheckResult result = checkFacilityDuplicates(
        name: 'Democare General Hospitl',
        type: FacilitySetupType.hospital,
        isActive: true,
        phone: '+256700000000',
        existing: existing,
      );

      expect(result.hasExactConflict, isFalse);
      expect(result.nonExactSimilarMatches, isNotEmpty);
      expect(
        result.nonExactSimilarMatches.first.score,
        greaterThanOrEqualTo(80),
      );
      expect(
        result.nonExactSimilarMatches.first.fieldComparisons,
        isNotEmpty,
      );
    });

    test('ignores excluded facility when editing', () {
      final FacilityDuplicateCheckResult result = checkFacilityDuplicates(
        name: 'DemoCare General Hospital',
        type: FacilitySetupType.hospital,
        isActive: true,
        existing: existing,
        excludeFacilityId: 'FAC0001',
      );

      expect(result.hasExactConflict, isFalse);
      expect(result.similarMatches, isEmpty);
    });

    test('ignores excluded facility by resource uuid alias', () {
      const List<FacilityProfile> withUuid = <FacilityProfile>[
        FacilityProfile(
          id: 'FAC0001',
          tenantId: 'TEN0001',
          name: 'DemoCare General Hospital',
          type: FacilitySetupType.hospital,
          resourceUuid: 'uuid-facility-1',
          displayId: 'FAC0001',
        ),
      ];

      final FacilityDuplicateCheckResult result = checkFacilityDuplicates(
        name: 'DemoCare General Hospital',
        type: FacilitySetupType.hospital,
        isActive: true,
        existing: withUuid,
        excludeFacilityId: 'uuid-facility-1',
      );

      expect(result.hasExactConflict, isFalse);
    });

    test('ignores excluded FacilityProfile across id aliases', () {
      const FacilityProfile editing = FacilityProfile(
        id: 'FAC0001',
        tenantId: 'TEN0001',
        name: 'DemoCare General Hospital',
        type: FacilitySetupType.hospital,
        resourceUuid: 'uuid-facility-1',
      );

      final FacilityDuplicateCheckResult result = checkFacilityDuplicates(
        name: 'DemoCare General Hospital',
        type: FacilitySetupType.hospital,
        isActive: true,
        existing: existing,
        excludeFacility: editing,
      );

      expect(result.hasExactConflict, isFalse);
    });
  });

  group('facility admin create/delete permissions', () {
    test('facility admin can manage but not create or delete', () {
      final AppAccessPolicy policy = AppAccessPolicy.fromSession(
        AuthSession(
          tokens: SessionTokens(accessToken: 't'),
          user: const AuthUserProfile(
            roles: <String>['FACILITY_ADMIN'],
            tenantId: 'TEN0001',
            facilityId: 'FAC0001',
          ),
        ),
      );

      expect(policy.canManageFacility(), isTrue);
      expect(policy.canCreateFacility(), isFalse);
      expect(policy.canDeleteFacility(), isFalse);
    });

    test('tenant admin can create and delete facilities', () {
      final AppAccessPolicy policy = AppAccessPolicy.fromSession(
        AuthSession(
          tokens: SessionTokens(accessToken: 't'),
          user: const AuthUserProfile(
            roles: <String>['TENANT_ADMIN'],
            tenantId: 'TEN0001',
          ),
        ),
      );

      expect(policy.canManageFacility(), isTrue);
      expect(policy.canCreateFacility(), isTrue);
      expect(policy.canDeleteFacility(), isTrue);
    });
  });
}

File: backend/src/tests/lib/facility/facility-similarity.test.js
const {
  checkFacilityDuplicates,
  compositeSimilarityScore,
  normalizeFacilityType
} = require('@lib/facility/facility-similarity');

describe('facility-similarity', () => {
  const existing = [
    {
      id: 'facility-1',
      human_friendly_id: 'FAC0001',
      tenant_id: 'TEN0001',
      name: 'DemoCare General Hospital',
      facility_type: 'HOSPITAL',
      is_active: true,
      contacts: [
        { contact_type: 'PHONE', value: '+256 700 000 000', is_primary: true },
        { contact_type: 'EMAIL', value: 'info@democare.test', is_primary: true }
      ],
      addresses: [
        {
          line1: '12 Kampala Road',
          city: 'Kampala',
          country: 'Uganda'
        }
      ]
    }
  ];

  it('normalizes facility type', () => {
    expect(normalizeFacilityType(' hospital ')).toBe('HOSPITAL');
  });

  it('weights name more heavily than status', () => {
    const nameHeavy = compositeSimilarityScore({
      nameScore: 100,
      statusScore: 0
    });
    const statusHeavy = compositeSimilarityScore({
      nameScore: 0,
      statusScore: 100
    });
    expect(nameHeavy).toBeGreaterThan(statusHeavy);
  });

  it('detects exact name conflict within tenant peers', () => {
    const result = checkFacilityDuplicates({
      name: 'DemoCare General Hospital',
      facilityType: 'HOSPITAL',
      isActive: true,
      existing
    });

    expect(result.exactNameConflict).toBe(true);
    expect(result.hasExactConflict).toBe(true);
    expect(result.similarMatches[0].isExact).toBe(true);
  });

  it('returns overridable similar matches for near names', () => {
    const result = checkFacilityDuplicates({
      name: 'Democare General Hospitl',
      facilityType: 'HOSPITAL',
      isActive: true,
      phone: '+256700000000',
      existing
    });

    expect(result.exactNameConflict).toBe(false);
    expect(result.overridableMatches.length).toBeGreaterThan(0);
    expect(result.overridableMatches[0].score).toBeGreaterThanOrEqual(80);
  });

  it('excludes the edited facility id', () => {
    const result = checkFacilityDuplicates({
      name: 'DemoCare General Hospital',
      facilityType: 'HOSPITAL',
      isActive: true,
      existing,
      excludeFacilityId: 'facility-1'
    });

    expect(result.exactNameConflict).toBe(false);
    expect(result.similarMatches).toHaveLength(0);
  });
});
